src/bridge/cockpitmetrics.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2014 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+
+#include "cockpitmetrics.h"
+#include "cockpitpcpmetrics.h"
+
+#include "common/cockpitjson.h"
+
+struct _CockpitMetricsPrivate {
+  guint timeout;
+  gint64 next;
+  gint64 interval;
+};
+
+G_DEFINE_ABSTRACT_TYPE (CockpitMetrics, cockpit_metrics, COCKPIT_TYPE_CHANNEL);
+
+static void
+cockpit_metrics_init (CockpitMetrics *self)
+{
+  self->priv = g_new0 (CockpitMetricsPrivate, 1);
+}
+
+static void
+cockpit_metrics_recv (CockpitChannel *channel,
+                      GBytes *message)
+{
+  g_warning ("received unexpected metrics1 payload");
+  cockpit_channel_close (channel, "protocol-error");
+}
+
+static void
+cockpit_metrics_close (CockpitChannel *channel,
+                       const gchar *problem)
+{
+  CockpitMetrics *self = COCKPIT_METRICS (channel);
+
+  if (self->priv->timeout)
+    {
+      g_source_remove (self->priv->timeout);
+      self->priv->timeout = 0;
+    }
+
+  COCKPIT_CHANNEL_CLASS (cockpit_metrics_parent_class)->close (channel, problem);
+}
+
+static void
+cockpit_metrics_dispose (GObject *object)
+{
+  CockpitMetrics *self = COCKPIT_METRICS (object);
+
+  if (self->priv->timeout)
+    {
+      g_source_remove (self->priv->timeout);
+      self->priv->timeout = 0;
+    }
+
+  G_OBJECT_CLASS (cockpit_metrics_parent_class)->dispose (object);
+}
+
+static void
+cockpit_metrics_finalize (GObject *object)
+{
+  CockpitMetrics *self = COCKPIT_METRICS (object);
+
+  g_free (self->priv);
+
+  G_OBJECT_CLASS (cockpit_metrics_parent_class)->finalize (object);
+}
+
+static void
+cockpit_metrics_class_init (CockpitMetricsClass *klass)
+{
+  CockpitChannelClass *channel_class = COCKPIT_CHANNEL_CLASS (klass);
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->dispose = cockpit_metrics_dispose;
+  object_class->finalize = cockpit_metrics_finalize;
+
+  channel_class->recv = cockpit_metrics_recv;
+  channel_class->close = cockpit_metrics_close;
+}
+
+static gboolean
+on_timeout_tick (gpointer data)
+{
+  CockpitMetrics *self = data;
+  CockpitMetricsClass *klass;
+  gint next_interval;
```

Comment:
This should be a gint64 and check for overflow below before passing to g_timeout_add()